### PR TITLE
Add event source for Amazon EventBridge

### LINKED ISSUE #633: Create source for Amazon EventBridge
_Originally posted by @life-learner1 in https://github.com/triggermesh/triggermesh/issues/632#issuecomment-1062692116_

---

I want to handle ECR event from EventBridge (CloudWatch Events).

image

That's the rule:

```json
{
  "detail-type": [
    "ECR Image Action"
  ],
  "source": [
    "aws.ecr"
  ],
  "detail": {
    "action-type": [
      "PUSH"
    ],
    "result": [
      "SUCCESS"
    ]
  }
}
```

### PR DESCRIPTION
**Pull request by antoineco**
Closes #633

Sources events from an Amazon EventBridge event bus by routing them over an Amazon SQS queue.
This flow is very similar to what we do in the S3 source.

The reconciliation is quite involved and requires:
1. Creating a _rule_ on the event bus
1. Creating a queue for routing events (unless provided by the user)
1. Applying an access policy to the queue to allow EventBridge to send events to it (unless provided by the user)
1. Setting the queue as a _target_ on the _rule_

---

### Source creation

```console
$ kubectl create -f /tmp/awseventbridgesource.yaml
awseventbridgesource.sources.triggermesh.io/sample created
```

```console
$ kubectl get awseventbridgesources.sources.triggermesh.io/sample
NAME     READY   REASON   QUEUE                                                                                SINK                                             AGE
sample   True             arn:aws:sqs:eu-west-1:123456789012:io_triggermesh_awseventbridgesources-4197625275   http://event-display.default.svc.cluster.local   4s
```

```console
$ kubectl describe awseventbridgesources.sources.triggermesh.io/sample
...
Events:
  Type    Reason           Age   From                             Message
  ----    ------           ----  ----                             -------
  Normal  FinalizerUpdate  9s    awseventbridgesource-controller  Updated "sample" finalizers
  Normal  QueueCreated     9s    awseventbridgesource-controller  Created SQS queue "https://sqs.eu-west-1.amazonaws.com/123456789012/io_triggermesh_awseventbridgesources-4197625275"
  Normal  Subscribed       8s    awseventbridgesource-controller  Configured rule "io_triggermesh_awseventbridgesources-4197625275" for event bus "arn:aws:events:eu-west-1:123456789012:event-bus/My-Test-Event-Bus"
  Normal  Subscribed       8s    awseventbridgesource-controller  Updated target for rule "io_triggermesh_awseventbridgesources-4197625275"
  Normal  CreateAdapter    8s    awseventbridgesource-controller  Created adapter Deployment "awseventbridgesource-sample"
```

### Managed AWS resources

image

image

image

### Data plane

Message sent

image

Message received

```jsonc
☁  cloudevents.Event
Context Attributes,
  specversion: 1.0
  source: aws:aws:events:eu-west-1:1234566789012:event-bus/My-Test-Event-Bus
  type: com.amazon.events.event
  id: 08ddd4c1-dbf6-c995-9126-602fa81d43eb
  time: 2022-05-23T18:12:54Z
  datacontenttype: application/json
Extensions,
  awseventsdetailtype: hello.v0
  awseventssource: com.example.test
Data,
  {
    "account": "123456789012",
    "detail": {
      "msg": "Hello, TriggerMesh!"
    },
    "detail-type": "hello.v0",
    "id": "08ddd4c1-dbf6-c995-9126-602fa81d43eb",
    "region": "eu-west-1",
    "resources": [],
    "source": "com.example.test",
    "time": "2022-05-23T18:12:54Z",
    "version": "0"
  }
```

### Source termination

```console
$ kubectl delete -f /tmp/awseventbridgesource.yaml
awseventbridgesource.sources.triggermesh.io "sample" deleted
```

```console
$ kubectl get events -w --sort-by='.lastTimestamp'
LAST SEEN   TYPE      REASON            OBJECT                        MESSAGE
8s          Normal    Unsubscribed      awseventbridgesource/sample   Removed rule "io_triggermesh_awseventbridgesources-4197625275" from event bus "arn:aws:events:eu-west-1:123456789012:event-bus/My-Test-Event-Bus"
8s          Normal    QueueDeleted      awseventbridgesource/sample   Deleted SQS queue "https://sqs.eu-west-1.amazonaws.com/123456789012/io_triggermesh_awseventbridgesources-4197625275"
8s          Normal    FinalizerUpdate   awseventbridgesource/sample   Updated "sample" finalizers
```

image

image

### CODE REVIEW

**Comment by sebgoa** [comment]
In your description above you seem to send an event of type `hello.v0` and that is also what is received. Therefore I presume that the source does not emit event of type `io.triggermesh.eventbridge` or something alike ? is that intended ?

**Comment by sebgoa** [comment]
apparently the event type is set from the `detail-type` of the eb event: https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-events.html

**Comment by antoineco** [comment]
Actually, now that you mention it, I think _I_ prefer that approach.
I'll push a change.

**Comment by antoineco** [comment]
@sebgoa done in [fbfb4b2e..b5f097c7](https://github.com/triggermesh/triggermesh/compare/fbfb4b2e932d73e28accd1d1d9205d4b4af40c01..b5f097c706c4beac19700662fa7e3ebdc66ae783).

I updated the sample event in the PR description. Please take a look.

Before you ask, the type is `com.amazon.events.event` and not `com.amazon.eventbridge.event` because we use the naming pattern `com.amazon.<service>.<type>` in all AWS sources, and the service is actually called "events", not "eventbridge", in Amazon's APIs (see the ARN, which also contains "events").

File: pkg/sources/adapter/awssqssource/process.go [hunk 1]
```diff
@@ -249,3 +279,93 @@ func isTestEventPayload(data map[string]interface{}) bool {
...
+		return false
+	}
+	_, ok := data["source"]
+	return ok
+}
+
+// makeEventBridgeEvent returns a CloudEvent for the given EventBridge event.
+func makeEventBridgeEvent(data map[string]interface{}) (*cloudevents.Event, error) {
+	event := cloudevents.NewEvent()
+	event.SetType(data["detail-type"].(string))
```
[the remark this answers is not in the document]
Comment (sebgoa) [reply pkg/sources/adapter/awssqssource/process.go]: ok go it. My comment was not on the stringification per se but the fact that the event type was going to be a sentence.

**Review by sebgoa — APPROVED** [review]
cursory look seems good but I have not tested it

**Comment by antoineco** [comment]
End-to-end tests added in [fae449ef..0a767c72](https://github.com/triggermesh/triggermesh/compare/fae449ef1b296f74ae6d682177794a2d64c6f1a7..0a767c727c01c0d374ecaab1d31453e67bc1e4ea).

Tested locally:

```
AWS EventBridge source
  a source watches an existing event bus
    when a message is sent to the event bus
        the source generates an event
/go/src/github.com/triggermesh/triggermesh/test/e2e/sources/awseventbridge/main.go:133
------------------------------

Ran 1 of 59 Specs in 22.476 seconds
SUCCESS! -- 1 Passed | 0 Failed | 1 Pending | 57 Skipped
```

**Pull request merged**